aten/src/ATen/FunctionalInverses.cpp

```diff
@@ -0,0 +1,191 @@
+
+#include <ATen/FunctionalInverses.h>
+
+#include <ATen/ATen.h>
+#include <ATen/ExpandUtils.h>
+namespace at {
+namespace functionalization {
+namespace impl {
+
+// This logic is similar to autograd code for view backwards calls.
+// We can't easily share it though, because (eventually) these functions
+// will all call `permute/unsqueeze_copy()` instead of `permute/unsqueeze`.
+
+Tensor permute_inverse(const Tensor& self, IntArrayRef dims) {
+  // invert the permutation
+  auto ndims = dims.size();
+  std::vector<int64_t> dims_(ndims);
+  for(const auto i : c10::irange(ndims)) {
+    dims_[at::maybe_wrap_dim(dims[i], ndims)] = i;
+  }
+  return self.permute(dims_);
+}
+
+Tensor unsqueeze_to(const Tensor & self, IntArrayRef sizes) {
+  auto result = self;
+
+  int64_t nDims = sizes.size();
+  for(const auto dim : c10::irange(nDims)) {
+    if (sizes[dim] == 1) {
+      result = result.unsqueeze(dim);
+    }
+  }
+  return result;
+}
+
+Tensor unsqueeze_to(const Tensor & self, int64_t dim, IntArrayRef sizes) {
+  dim = at::maybe_wrap_dim(dim, sizes.size());
+  // in NumPy it's not an error to unsqueeze a scalar, but we still need to avoided
+  // unsqueezing in the backward.
+  if (sizes.size() > 0 && sizes[dim] == 1) {
+    return self.unsqueeze(dim);
+  }
+  return self;
+}
+
+
+// ----------------------------------------------------------
+// Implementations of each view_inverse() function are below.
+// One of these needs to be implemented for every existing non-composite view operator.
+// The codegen automatically generates the corresponding function declaration.
+// ----------------------------------------------------------
```

Comment:
Maybe worth commenting on what it means when base is not used?

Reply:
I'm curious if you are accidentally retaining a reference to base, even when the inverse doesn't need it!

Reply:
Definitely deserves a comment.

I left the reference on purpose, mostly because figuring out when to remove it in the codegen is annoying (I think it would involve hardcoding the slice/select/diagonal names somewhere).

In terms of clarity, that's probably confusing and deserves a comment.
In terms of perf, I figured that the impact of passing in an extra reference param unnecessarily isn't too bad, compared to some other lower hanging perf fruit.